File: apps/api/src/services/analysis/weibull.ts
/**
 * Area-mean Weibull A/k over an AOI, read from the local GWA combined-Weibull
 * country COGs (250 m grid, float32, NaN nodata, EPSG:4326 — VERIFIED.md §2).
 *
 * The COGs are fetched once by `scripts/fetch-weibull-cogs.ts` into
 * WEIBULL_COG_DIR. When they are absent or unreadable this module degrades
 * gracefully: `aoiWeibullMeans` resolves null (logged once at first use, not
 * per call) and the resource section ships without a distribution.
 *
 * Point-in-ring note: mask.ts (built in parallel) targets stitched
 * web-mercator LayerPatch grids at ANALYSIS_ZOOM (see types.ts AoiMask) — its
 * API does not fit the COGs' plain lon/lat grid, so a tiny local ray-cast
 * helper is the deliberate DRY tradeoff here.
 */

import { fromFile, type GeoTIFF, type GeoTIFFImage } from "geotiff";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import {
  WEIBULL_A_FILE,
  WEIBULL_COG_DIR,
  WEIBULL_K_FILE,
} from "./constants";

// ── Paths (shared with scripts/fetch-weibull-cogs.ts and the tests) ────────

const MODULE_DIR = dirname(fileURLToPath(import.meta.url));
/** apps/api root, independent of process.cwd(). */
const API_ROOT = join(MODULE_DIR, "..", "..", "..");

export const WEIBULL_COG_PATHS = {
  a: join(API_ROOT, WEIBULL_COG_DIR, WEIBULL_A_FILE),
  k: join(API_ROOT, WEIBULL_COG_DIR, WEIBULL_K_FILE),
} as const;

// ── Geotransform / pixel-window math (pure, unit-tested) ───────────────────

/** North-up affine geotransform of a COG, read from its image metadata. */
export interface GeoTransform {
  /** Lon of the LEFT edge of pixel column 0 (degrees). */
  originX: number;
  /** Lat of the TOP edge of pixel row 0 (degrees). */
  originY: number;
  /** Degrees per pixel eastward (> 0). */
  pixelWidth: number;
  /** Degrees per pixel southward (< 0 for north-up rasters). */
  pixelHeight: number;
  widthPx: number;
  heightPx: number;
}

/** Half-open pixel window [x0, x1) × [y0, y1) in full-image coordinates. */
export interface PixelWindow {
  x0: number;
  y0: number;
  x1: number;
  y1: number;
}

/**
 * Defensive cap on a single window read. The 2,500 km² AOI cap is ~40k pixels
 * on the 250 m grid; anything near this limit signals a bbox bug upstream.
 */
export const MAX_WINDOW_PIXELS = 4_000_000;

const TRANSFORM_MATCH_EPSILON_DEG = 1e-9;

/**
 * Pixel window covering a lon/lat bbox [W, S, E, N] at native resolution,
 * expanded outward to whole pixels and clamped to the image. Returns null
 * when the bbox does not overlap the raster; throws when the window would
 * exceed MAX_WINDOW_PIXELS (upstream bbox bug, never a data condition).
 */
export function bboxPixelWindow(
  transform: GeoTransform,
  bbox: readonly [number, number, number, number],
): PixelWindow | null {
  const { originX, originY, pixelWidth, pixelHeight, widthPx, heightPx } =
    transform;
  if (!(pixelWidth > 0) || !(pixelHeight < 0)) {
    throw new Error(
      `bboxPixelWindow: expected north-up transform (pixelWidth>0, pixelHeight<0), got ${pixelWidth}/${pixelHeight}`,
    );
  }
  const [west, south, east, north] = bbox;
  const x0 = Math.max(0, Math.floor((west - originX) / pixelWidth));
  const x1 = Math.min(widthPx, Math.ceil((east - originX) / pixelWidth));
  // North (larger lat) maps to the SMALLER row index because pixelHeight < 0.
  const y0 = Math.max(0, Math.floor((north - originY) / pixelHeight));
  const y1 = Math.min(heightPx, Math.ceil((south - originY) / pixelHeight));
  if (x1 <= x0 || y1 <= y0) return null;

  const pixelCount = (x1 - x0) * (y1 - y0);
  if (pixelCount > MAX_WINDOW_PIXELS) {
    throw new Error(
      `bboxPixelWindow: window of ${pixelCount} px exceeds cap ${MAX_WINDOW_PIXELS} — bbox bug upstream (bbox=${bbox.join(",")})`,
    );
  }
  return { x0, y0, x1, y1 };
}

/** Lon/lat of the CENTER of pixel (col, row) in full-image coordinates. */
export function pixelCenterLngLat(
  transform: GeoTransform,
  col: number,
  row: number,
): [number, number] {
  return [
    transform.originX + (col + 0.5) * transform.pixelWidth,
    transform.originY + (row + 0.5) * transform.pixelHeight,
  ];
}

// ── Point-in-ring (even-odd ray cast) ───────────────────────────────────────

/** True when [lon, lat] falls inside the closed lon/lat ring (even-odd). */
export function isInsideRing(
  lon: number,
  lat: number,
  ring: readonly (readonly [number, number])[],
): boolean {
  let isInside = false;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const [xi, yi] = ring[i]!;
    const [xj, yj] = ring[j]!;
    const crossesRay =
      yi > lat !== yj > lat &&
      lon < ((xj - xi) * (lat - yi)) / (yj - yi) + xi;
    if (crossesRay) isInside = !isInside;
  }
  return isInside;
}

// ── Gamma (Lanczos g=7) — for the Weibull mean A·Γ(1+1/k) ──────────────────

const LANCZOS_G = 7;
const LANCZOS_COEFFICIENTS = [
  0.99999999999980993, 676.5203681218851, -1259.1392167224028,
  771.32342877765313, -176.61502916214059, 12.507343278686905,
  -0.13857109526572012, 9.9843695780195716e-6, 1.5056327351493116e-7,
] as const;

/** Gamma function via the Lanczos approximation (g=7, 9 coefficients). */
export function gammaFn(x: number): number {
  if (x < 0.5) {
    // Reflection formula for the left half-plane.
    return Math.PI / (Math.sin(Math.PI * x) * gammaFn(1 - x));
  }
  const z = x - 1;
  let sum = LANCZOS_COEFFICIENTS[0];
  for (let i = 1; i < LANCZOS_COEFFICIENTS.length; i++) {
    sum += LANCZOS_COEFFICIENTS[i]! / (z + i);
  }
  const t = z + LANCZOS_G + 0.5;
  return Math.sqrt(2 * Math.PI) * t ** (z + 0.5) * Math.exp(-t) * sum;
}

// ── COG handles (opened once, lazily; null = degraded mode) ────────────────

interface OpenCog {
  /** Kept so the OS file descriptor can be closed (tests, graceful shutdown). */
  tiff: GeoTIFF;
  image: GeoTIFFImage;
  transform: GeoTransform;
}

interface WeibullCogs {
  a: OpenCog;
  k: OpenCog;
}

let cogsPromise: Promise<WeibullCogs | null> | null = null;

function loadWeibullCogs(): Promise<WeibullCogs | null> {
  // The cached promise must NEVER be rejected — a rejected singleton would
  // throw on every subsequent call instead of degrading to null.
  cogsPromise ??= openBothCogs().catch((err) => {
    console.error(
      "[analysis/weibull] unexpected COG load failure — Weibull means degrade to null:",
      err,
    );
    return null;
  });
  return cogsPromise;
}

/** Best-effort close of COG file descriptors; logs, never throws. */
async function closeTiffHandles(tiffs: readonly GeoTIFF[]): Promise<void> {
  for (const tiff of tiffs) {
    try {
      const pending = tiff.close(); // false when not file-backed
      if (pending !== false) await pending;
    } catch (err) {
      console.warn("[analysis/weibull] failed to close COG handle:", err);
    }
  }
}

async function openCog(path: string): Promise<OpenCog> {
  const tiff = await fromFile(path);
  try {
    const image = await tiff.getImage(0); // index 0 = full-resolution IFD
    const transform: GeoTransform = {
      originX: image.getOrigin()[0] ?? Number.NaN,
      originY: image.getOrigin()[1] ?? Number.NaN,
      pixelWidth: image.getResolution()[0] ?? Number.NaN,
      pixelHeight: image.getResolution()[1] ?? Number.NaN,
      widthPx: image.getWidth(),
      heightPx: image.getHeight(),
    };
    const fields = [
      transform.originX,
      transform.originY,
      transform.pixelWidth,
      transform.pixelHeight,
    ];
    if (!fields.every(Number.isFinite)) {
      throw new Error(`Weibull COG ${path}: malformed geotransform`);
    }
    return { tiff, image, transform };
  } catch (err) {
    // Never strand the fd when metadata reads fail mid-open.
    await closeTiffHandles([tiff]);
    throw err;
  }
}

const isFulfilled = <T>(
  result: PromiseSettledResult<T>,
): result is PromiseFulfilledResult<T> => result.status === "fulfilled";

/** A and k must share one grid or pixel pairing would silently misalign. */
function isSameGrid(a: GeoTransform, b: GeoTransform): boolean {
  return (
    a.widthPx === b.widthPx &&
    a.heightPx === b.heightPx &&
    Math.abs(a.originX - b.originX) < TRANSFORM_MATCH_EPSILON_DEG &&
    Math.abs(a.originY - b.originY) < TRANSFORM_MATCH_EPSILON_DEG &&
    Math.abs(a.pixelWidth - b.pixelWidth) < TRANSFORM_MATCH_EPSILON_DEG &&
    Math.abs(a.pixelHeight - b.pixelHeight) < TRANSFORM_MATCH_EPSILON_DEG
  );
}

async function openBothCogs(): Promise<WeibullCogs | null> {
  const [hasA, hasK] = await Promise.all([
    Bun.file(WEIBULL_COG_PATHS.a).exists(),
    Bun.file(WEIBULL_COG_PATHS.k).exists(),
  ]);
  if (!hasA || !hasK) {
    console.warn(
      `[analysis/weibull] COG(s) missing under ${join(API_ROOT, WEIBULL_COG_DIR)} ` +
        `(A: ${hasA}, k: ${hasK}) — run \`bun scripts/fetch-weibull-cogs.ts\`. ` +
        `Weibull means degrade to null.`,
    );
    return null;
  }
  // allSettled so a half-open pair can be closed — Promise.all would strand
  // the fd of whichever COG opened successfully when the other one failed.
  const settled = await Promise.allSettled([
    openCog(WEIBULL_COG_PATHS.a),
    openCog(WEIBULL_COG_PATHS.k),
  ]);
  const opened = settled.filter(isFulfilled).map((result) => result.value);
  const failed = settled.find(
    (result): result is PromiseRejectedResult => result.status === "rejected",
  );
  if (failed !== undefined || opened.length !== 2) {
    await closeTiffHandles(opened.map((cog) => cog.tiff));
    console.error(
      "[analysis/weibull] failed to open COGs — Weibull means degrade to null:",
      failed?.reason,
    );
    return null;
  }
  const [a, k] = opened as [OpenCog, OpenCog];
  if (!isSameGrid(a.transform, k.transform)) {
    await closeTiffHandles([a.tiff, k.tiff]);
    console.error(
      "[analysis/weibull] A and k COGs have mismatched grids — refusing to pair pixels; Weibull means degrade to null.",
    );
    return null;
  }
  return { a, k };
}

/**
 * Close both COG file descriptors (if open) and clear the lazy singleton so
 * the next aoiWeibullMeans call re-opens from disk. For tests (isolation
 * between degraded/live paths) and graceful-shutdown hooks — production
 * request paths never call this.
 */
export async function resetWeibullCogs(): Promise<void> {
  const pending = cogsPromise;
  cogsPromise = null;
  if (pending === null) return;
  // openBothCogs resolves null on every failure path (never rejects), but a
  // defensive catch keeps reset itself unfailable.
  const cogs = await pending.catch((err) => {
    console.warn("[analysis/weibull] reset found a rejected COG load:", err);
    return null;
  });
  if (cogs !== null) {
    await closeTiffHandles([cogs.a.tiff, cogs.k.tiff]);
  }
}

// ── Public API ──────────────────────────────────────────────────────────────

async function readWindowBand(
  image: GeoTIFFImage,
  window: PixelWindow,
): Promise<Float32Array> {
  const rasters = await image.readRasters({
    window: [window.x0, window.y0, window.x1, window.y1],
  });
  const band = rasters[0];
  if (!(band instanceof Float32Array)) {
    throw new Error("Weibull COG read returned a non-float32 band");
  }
  return band;
}

/**
 * Area-mean Weibull A and k over the AOI: native-resolution window read
 * covering `bbox`, then the mean of every finite, positive A/k pixel pair
 * whose center falls inside `ring`.
 *
 * Returns null when the COGs are unavailable (degraded mode), the bbox
 * misses the raster, the read fails, or zero in-ring pixels are valid.
 * Throws only for the MAX_WINDOW_PIXELS cap (upstream bbox bug).
 */
export async function aoiWeibullMeans(
  bbox: readonly [number, number, number, number],
  ring: readonly (readonly [number, number])[],
): Promise<{ A: number; k: number } | null> {
  const cogs = await loadWeibullCogs();
  if (!cogs) return null;

  // Grids verified identical at load time — one window serves both reads.
  const window = bboxPixelWindow(cogs.a.transform, bbox);
  if (!window) return null;

  let aBand: Float32Array;
  let kBand: Float32Array;
  try {
    [aBand, kBand] = await Promise.all([
      readWindowBand(cogs.a.image, window),
      readWindowBand(cogs.k.image, window),
    ]);
  } catch (err) {
    console.error("[analysis/weibull] window read failed — returning null:", err);
    return null;
  }

  const windowWidth = window.x1 - window.x0;
  let sumA = 0;
  let sumK = 0;
  let insideCount = 0;
  for (let row = window.y0; row < window.y1; row++) {
    for (let col = window.x0; col < window.x1; col++) {
      const i = (row - window.y0) * windowWidth + (col - window.x0);
      const aValue = aBand[i] ?? Number.NaN;
      const kValue = kBand[i] ?? Number.NaN;
      const isValidPair =
        Number.isFinite(aValue) && Number.isFinite(kValue) && aValue > 0 && kValue > 0;
      if (!isValidPair) continue;
      const [lon, lat] = pixelCenterLngLat(cogs.a.transform, col, row);
      if (!isInsideRing(lon, lat, ring)) continue;
      sumA += aValue;
      sumK += kValue;
      insideCount += 1;
    }
  }

  if (insideCount === 0) return null;
  return { A: sumA / insideCount, k: sumK / insideCount };
}

File: services/site-analysis/app/engine/weibull.py
"""Area-mean Weibull A/k over an AOI, read from the local GWA combined-Weibull
country COGs (250 m grid, float32, NaN nodata, EPSG:4326-by-assumption —
VERIFIED.md §2).

Verbatim port of apps/api/src/services/analysis/weibull.ts. The COGs are fetched
once by ``scripts/fetch-weibull-cogs.ts`` into WEIBULL_COG_DIR; here they resolve
to ``config.WEIBULL_A_PATH`` / ``config.WEIBULL_K_PATH`` under DATA_DIR. When they
are absent or unreadable this module degrades gracefully: ``aoi_weibull_means``
returns None (logged once at first use, not per call) and the resource section
ships without a distribution.

Point-in-ring note: mask.py (foundation, wave 1) targets stitched web-mercator
LayerPatch grids at ANALYSIS_ZOOM (see types.py AoiMask) — its API does not fit
the COGs' plain lon/lat grid, so a tiny LOCAL ray-cast helper (``is_inside_ring``)
is the deliberate DRY tradeoff here. This is the documented dual ray-cast.

The Weibull COGs report ``crs=None``; per the migration contract we treat
``ds.transform`` (an Affine) as a lon/lat DEGREE grid by assumption and do NOT
rely on an embedded CRS. The TS reads ``image.getOrigin()`` / ``getResolution()``
from the GeoTIFF header — the rasterio equivalents are:

    originX     = transform.c   (lon of LEFT edge of pixel column 0)
    originY     = transform.f   (lat of TOP edge of pixel row 0)
    pixelWidth  = transform.a   (degrees/px eastward, > 0)
    pixelHeight = transform.e   (degrees/px southward, < 0 for north-up)
"""
from __future__ import annotations

import logging
import math
import os
from dataclasses import dataclass
from typing import Optional, Sequence

import numpy as np

from app import config

logger = logging.getLogger(__name__)


# ── Paths (shared with scripts/fetch-weibull-cogs.ts and the tests) ────────
#
# The TS resolves these relative to import.meta.url; the Python service pins them
# in config (DATA_DIR / "gwa" / <file>) so monkeypatched paths are honoured in
# tests exactly like indiacdf's INDIA_CDF_PATH. Read through ``config`` at use
# time — never bind the Path at import.

# ── Geotransform / pixel-window math (pure, unit-tested) ───────────────────


@dataclass(frozen=True)
class GeoTransform:
    """North-up affine geotransform of a COG, read from its image metadata."""

    # Lon of the LEFT edge of pixel column 0 (degrees).
    origin_x: float
    # Lat of the TOP edge of pixel row 0 (degrees).
    origin_y: float
    # Degrees per pixel eastward (> 0).
    pixel_width: float
    # Degrees per pixel southward (< 0 for north-up rasters).
    pixel_height: float
    width_px: int
    height_px: int


@dataclass(frozen=True)
class PixelWindow:
    """Half-open pixel window [x0, x1) × [y0, y1) in full-image coordinates."""

    x0: int
    y0: int
    x1: int
    y1: int


# Defensive cap on a single window read. The 2,500 km² AOI cap is ~40k pixels on
# the 250 m grid; anything near this limit signals a bbox bug upstream.
MAX_WINDOW_PIXELS = 4_000_000

TRANSFORM_MATCH_EPSILON_DEG = 1e-9


def bbox_pixel_window(
    transform: GeoTransform,
    bbox: Sequence[float],
) -> Optional[PixelWindow]:
    """Pixel window covering a lon/lat bbox ``[W, S, E, N]`` at native resolution,
    expanded outward to whole pixels and clamped to the image. Returns None when
    the bbox does not overlap the raster; throws when the window would exceed
    MAX_WINDOW_PIXELS (upstream bbox bug, never a data condition).
    (weibull.ts:74-100)
    """
    origin_x = transform.origin_x
    origin_y = transform.origin_y
    pixel_width = transform.pixel_width
    pixel_height = transform.pixel_height
    width_px = transform.width_px
    height_px = transform.height_px
    if not (pixel_width > 0) or not (pixel_height < 0):
        raise ValueError(
            "bboxPixelWindow: expected north-up transform "
            f"(pixelWidth>0, pixelHeight<0), got {pixel_width}/{pixel_height}"
        )
    west, south, east, north = bbox[0], bbox[1], bbox[2], bbox[3]
    x0 = max(0, math.floor((west - origin_x) / pixel_width))
    x1 = min(width_px, math.ceil((east - origin_x) / pixel_width))
    # North (larger lat) maps to the SMALLER row index because pixelHeight < 0.
    y0 = max(0, math.floor((north - origin_y) / pixel_height))
    y1 = min(height_px, math.ceil((south - origin_y) / pixel_height))
    if x1 <= x0 or y1 <= y0:
        return None

    pixel_count = (x1 - x0) * (y1 - y0)
    if pixel_count > MAX_WINDOW_PIXELS:
        bbox_str = ",".join(str(v) for v in bbox)
        raise ValueError(
            f"bboxPixelWindow: window of {pixel_count} px exceeds cap "
            f"{MAX_WINDOW_PIXELS} — bbox bug upstream (bbox={bbox_str})"
        )
    return PixelWindow(x0=x0, y0=y0, x1=x1, y1=y1)


def pixel_center_lng_lat(
    transform: GeoTransform,
    col: int,
    row: int,
) -> tuple[float, float]:
    """Lon/lat of the CENTER of pixel (col, row) in full-image coordinates.
    (weibull.ts:103-112)
    """
    return (
        transform.origin_x + (col + 0.5) * transform.pixel_width,
        transform.origin_y + (row + 0.5) * transform.pixel_height,
    )


# ── Point-in-ring (even-odd ray cast) ───────────────────────────────────────


def is_inside_ring(
    lon: float,
    lat: float,
    ring: Sequence[Sequence[float]],
) -> bool:
    """True when ``[lon, lat]`` falls inside the closed lon/lat ring (even-odd).
    (weibull.ts:117-132)
    """
    is_inside = False
    j = len(ring) - 1
    for i in range(len(ring)):
        xi, yi = ring[i][0], ring[i][1]
        xj, yj = ring[j][0], ring[j][1]
        crosses_ray = (yi > lat) != (yj > lat) and (
            lon < (xj - xi) * (lat - yi) / (yj - yi) + xi
        )
        if crosses_ray:
            is_inside = not is_inside
        j = i
    return is_inside


# ── Gamma (Lanczos g=7) — for the Weibull mean A·Γ(1+1/k) ──────────────────

LANCZOS_G = 7
LANCZOS_COEFFICIENTS = (
    0.99999999999980993,
    676.5203681218851,
    -1259.1392167224028,
    771.32342877765313,
    -176.61502916214059,
    12.507343278686905,
    -0.13857109526572012,
    9.9843695780195716e-6,
    1.5056327351493116e-7,
)


def gamma_fn(x: float) -> float:
    """Gamma function via the Lanczos approximation (g=7, 9 coefficients).
    (weibull.ts:144-156)

    NOTE: present for completeness — it is NOT invoked by ``aoi_weibull_means``
    at request time (the weibull mean A·Γ(1+1/k) is computed downstream / in the
    live test, never here).
    """
    if x < 0.5:
        # Reflection formula for the left half-plane.
        return math.pi / (math.sin(math.pi * x) * gamma_fn(1 - x))
    z = x - 1
    total = LANCZOS_COEFFICIENTS[0]
    for i in range(1, len(LANCZOS_COEFFICIENTS)):
        total += LANCZOS_COEFFICIENTS[i] / (z + i)
    t = z + LANCZOS_G + 0.5
    return math.sqrt(2 * math.pi) * t ** (z + 0.5) * math.exp(-t) * total


# ── COG handles (opened once, lazily; None = degraded mode) ────────────────


@dataclass(frozen=True)
class OpenCog:
    """An opened COG dataset plus its parsed north-up geotransform. ``dataset`` is
    a rasterio DatasetReader (the analogue of geotiff's GeoTIFF/GeoTIFFImage pair);
    keeping it lets the OS file descriptor be closed (tests, graceful shutdown)."""

    dataset: object  # rasterio.io.DatasetReader
    transform: GeoTransform


@dataclass(frozen=True)
class WeibullCogs:
    a: OpenCog
    k: OpenCog


# undefined (not loaded yet) -> _NOT_LOADED; None = attempted and unavailable;
# WeibullCogs = open. Mirrors the TS lazy promise singleton (which never rejects).
_NOT_LOADED = object()
_cogs_cache: object = _NOT_LOADED


def _load_weibull_cogs() -> Optional[WeibullCogs]:
    """Lazy singleton load. The cache must NEVER carry a thrown error — a thrown
    singleton would raise on every subsequent call instead of degrading to None.
    (weibull.ts:174-185 loadWeibullCogs)
    """
    global _cogs_cache
    if _cogs_cache is not _NOT_LOADED:
        return _cogs_cache  # type: ignore[return-value]
    try:
        _cogs_cache = _open_both_cogs()
    except Exception as err:  # noqa: BLE001 — mirror the TS .catch, never throws
        logger.error(
            "[analysis/weibull] unexpected COG load failure — "
            "Weibull means degrade to null: %s",
            err,
        )
        _cogs_cache = None
    return _cogs_cache  # type: ignore[return-value]


def _close_dataset_handles(datasets: Sequence[object]) -> None:
    """Best-effort close of COG file descriptors; logs, never throws.
    (weibull.ts:188-197 closeTiffHandles)
    """
    for dataset in datasets:
        try:
            dataset.close()  # type: ignore[attr-defined]
        except Exception as err:  # noqa: BLE001
            logger.warning("[analysis/weibull] failed to close COG handle: %s", err)


def _open_cog(path) -> OpenCog:
    """Open one COG and parse its north-up geotransform from ``ds.transform``
    (treated as a lon/lat degree grid by assumption — crs is None). Never strands
    the fd when metadata reads fail mid-open. (weibull.ts:199-226 openCog)
    """
    import rasterio

    dataset = rasterio.open(path)
    try:
        affine = dataset.transform
        transform = GeoTransform(
            origin_x=affine.c,
            origin_y=affine.f,
            pixel_width=affine.a,
            pixel_height=affine.e,
            width_px=dataset.width,
            height_px=dataset.height,
        )
        fields = [
            transform.origin_x,
            transform.origin_y,
            transform.pixel_width,
            transform.pixel_height,
        ]
        if not all(isinstance(v, (int, float)) and math.isfinite(v) for v in fields):
            raise ValueError(f"Weibull COG {path}: malformed geotransform")
        return OpenCog(dataset=dataset, transform=transform)
    except Exception:
        # Never strand the fd when metadata reads fail mid-open.
        _close_dataset_handles([dataset])
        raise


def _is_same_grid(a: GeoTransform, b: GeoTransform) -> bool:
    """A and k must share one grid or pixel pairing would silently misalign.
    (weibull.ts:233-242 isSameGrid)
    """
    return (
        a.width_px == b.width_px
        and a.height_px == b.height_px
        and abs(a.origin_x - b.origin_x) < TRANSFORM_MATCH_EPSILON_DEG
        and abs(a.origin_y - b.origin_y) < TRANSFORM_MATCH_EPSILON_DEG
        and abs(a.pixel_width - b.pixel_width) < TRANSFORM_MATCH_EPSILON_DEG
        and abs(a.pixel_height - b.pixel_height) < TRANSFORM_MATCH_EPSILON_DEG
    )


def _open_both_cogs() -> Optional[WeibullCogs]:
    """Open both COGs, refusing to pair pixels unless the grids match. Closes any
    half-open fd on every failure path and degrades to None. (weibull.ts:244-284)
    """
    a_path = config.WEIBULL_A_PATH
    k_path = config.WEIBULL_K_PATH
    has_a = os.path.exists(a_path)
    has_k = os.path.exists(k_path)
    if not has_a or not has_k:
        logger.warning(
            "[analysis/weibull] COG(s) missing under %s "
            "(A: %s, k: %s) — run `bun scripts/fetch-weibull-cogs.ts`. "
            "Weibull means degrade to null.",
            os.path.dirname(str(a_path)),
            has_a,
            has_k,
        )
        return None
    # Open each independently so a half-open pair can be closed — opening both in
    # one try and letting an exception escape would strand the fd of whichever COG
    # opened successfully when the other one failed.
    opened: list[OpenCog] = []
    failed_reason: Optional[BaseException] = None
    for path in (a_path, k_path):
        try:
            opened.append(_open_cog(path))
        except Exception as err:  # noqa: BLE001
            failed_reason = err
    if failed_reason is not None or len(opened) != 2:
        _close_dataset_handles([cog.dataset for cog in opened])
        logger.error(
            "[analysis/weibull] failed to open COGs — "
            "Weibull means degrade to null: %s",
            failed_reason,
        )
        return None
    a, k = opened[0], opened[1]
    if not _is_same_grid(a.transform, k.transform):
        _close_dataset_handles([a.dataset, k.dataset])
        logger.error(
            "[analysis/weibull] A and k COGs have mismatched grids — "
            "refusing to pair pixels; Weibull means degrade to null."
        )
        return None
    return WeibullCogs(a=a, k=k)


def reset_weibull_cogs() -> None:
    """Close both COG file descriptors (if open) and clear the lazy singleton so
    the next ``aoi_weibull_means`` call re-opens from disk. For tests (isolation
    between degraded/live paths) and graceful-shutdown hooks — production request
    paths never call this. (weibull.ts:292-305 resetWeibullCogs)
    """
    global _cogs_cache
    pending = _cogs_cache
    _cogs_cache = _NOT_LOADED
    if pending is _NOT_LOADED or pending is None:
        return
    cogs: Optional[WeibullCogs] = pending  # type: ignore[assignment]
    if cogs is not None:
        _close_dataset_handles([cogs.a.dataset, cogs.k.dataset])


# ── Public API ──────────────────────────────────────────────────────────────


def _read_window_band(dataset, window: PixelWindow) -> np.ndarray:
    """Read band 0 over the half-open pixel window, row-major, as float32 — the
    rasterio analogue of geotiff ``readRasters({window:[x0,y0,x1,y1]})``.
    (weibull.ts:309-321 readWindowBand)
    """
    from rasterio.windows import Window

    band = dataset.read(
        1,
        window=Window(
            window.x0,
            window.y0,
            window.x1 - window.x0,
            window.y1 - window.y0,
        ),
    )
    if band.dtype != np.float32:
        raise ValueError("Weibull COG read returned a non-float32 band")
    return band


def aoi_weibull_means(
    bbox: Sequence[float],
    ring: Sequence[Sequence[float]],
) -> Optional[dict]:
    """Area-mean Weibull A and k over the AOI: native-resolution window read
    covering ``bbox``, then the mean of every finite, positive A/k pixel pair
    whose center falls inside ``ring``. A and k are returned UNROUNDED.

    Returns None when the COGs are unavailable (degraded mode), the bbox misses
    the raster, the read fails, or zero in-ring pixels are valid. Throws only for
    the MAX_WINDOW_PIXELS cap (upstream bbox bug). (weibull.ts:332-377)
    """
    cogs = _load_weibull_cogs()
    if not cogs:
        return None

    # Grids verified identical at load time — one window serves both reads.
    window = bbox_pixel_window(cogs.a.transform, bbox)
    if not window:
        return None

    try:
        a_band = _read_window_band(cogs.a.dataset, window)
        k_band = _read_window_band(cogs.k.dataset, window)
    except Exception as err:  # noqa: BLE001
        logger.error(
            "[analysis/weibull] window read failed — returning null: %s", err
        )
        return None

    # rasterio returns 2-D (rows, cols); index by (row-y0, col-x0) which is the
    # same row-major layout the TS reads as a flat Float32Array.
    sum_a = 0.0
    sum_k = 0.0
    inside_count = 0
    for row in range(window.y0, window.y1):
        for col in range(window.x0, window.x1):
            a_value = float(a_band[row - window.y0, col - window.x0])
            k_value = float(k_band[row - window.y0, col - window.x0])
            is_valid_pair = (
                math.isfinite(a_value)
                and math.isfinite(k_value)
                and a_value > 0
                and k_value > 0
            )
            if not is_valid_pair:
                continue
            lon, lat = pixel_center_lng_lat(cogs.a.transform, col, row)
            if not is_inside_ring(lon, lat, ring):
                continue
            sum_a += a_value
            sum_k += k_value
            inside_count += 1

    if inside_count == 0:
        return None
    return {"A": sum_a / inside_count, "k": sum_k / inside_count}
